Load Sample Data into MongoDB via a NPM Command

anshgoyalevil opened the pull request:
<!--
This section can be deleted after reading.

We employ the following branching strategy to simplify the development process and to ensure that only stable code is pushed to the `master` branch:

- `develop`: For unstable code: New features and bug fixes.
- `alpha-x.x.x`: For stability testing: Only bug fixes accepted.
- `master`: Where the stable production ready code lies. Only security related bugs.

-->

<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

**What kind of change does this PR introduce?**
This PR adds a dynamic NPM command which parses the command line arguments and executes a data loader with those arguments.
<!-- E.g. a bugfix, feature, refactoring, etc… -->

**Issue Number:**

Fixes #1293 

**Did you add tests for your changes?**
No
<!--Yes or No. Note: Add unit tests or automation tests for your code.-->

**Snapshots/Videos:**
![image](https://user-images.githubusercontent.com/94157520/231945728-c975a3e4-315d-49bc-902a-7c9f63660048.png)
![image](https://user-images.githubusercontent.com/94157520/231945832-c9724ae8-7a5a-45dd-b37b-8930ea25b369.png)

<!--Add snapshots or videos wherever possible.-->

**If relevant, did you update the documentation?**
Yes
<!--Add link to Talawa-Docs.-->

**Summary**
This PR adds a dynamic NPM command which parses the command line arguments and executes a data loader with those arguments.
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**Does this PR introduce a breaking change?**
No
<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

**Other information**

## Syntax:

```npm run import:sample-data -- [args]```
You can pass the following arguments while running this script.
- ```--format```: Cleans the database before import. **Add this flag with caution. It will delete all of the existing data inside the talawa database.**
- ```--items=```: Specify the items to add.
    - Following ```items``` can be specified, separated with a comma ```,```
        - ```users```: For users collection
        - ```organizations```: For organizations collection
        - ```events```: For events collection
        - ```posts```: For posts collection
## Examples:
- ```npm run import:sample-data```: This command will import the complete sample database without removing the existing data.
- ```npm run import:sample-data -- --format```: This command will import the complete sample database after removing the existing data.
- ```npm run import:sample-data -- --format --items=users,organizations```: This command will import the sample ```users``` and ```organizations``` collections after cleaning the existing data.
- ```npm run import:sample-data --  --items=users,organizations```: This command will import the sample ```users``` and ```organizations``` collections without cleaning the existing data.

<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**
Yes
<!--Yes or No-->

anshgoyalevil commented:
@palisadoes 

Please review. I have updated the ```installation.md``` also.

For the automatic database cleanup and import, There are cloud functions available which trigger a script at regular interval of time.

For example, in Azure, we have [WebJobs](https://learn.microsoft.com/en-us/azure/app-service/webjobs-create)
In AWS, we have [AWS Lambda with CloudWatch](https://www.freecodecamp.org/news/using-lambda-functions-as-cronjobs/)

I am researching a better way for automation in cloud.

The

EshaanAgg commented:
@anshgoyalevil Don't you think we should add a couple of more database objects to the samples?  One sample object per type itsn't really great for testing out the functionality of the app in my opinion.

anshgoyalevil commented:
@EshaanAgg 

I was also thinking the same.

What do you suggest? Adding more users, orgs, or any other collection?

anshgoyalevil commented:
Alright, thanks for suggestions.

EshaanAgg commented on src/utilities/loadSampleData.ts:
```diff
@@ -0,0 +1,98 @@
+
+yargs.option("f", {
+  alias: "format",
+  describe: "Whether to format the database before insertion",
```
`Formats all the collections present in the database before the insertion of objects. [WARNING] Use carefully.`

This is a better description is my opinion.

anshgoyalevil replied:
Yes. Would change it.

EshaanAgg commented:
@anshgoyalevil Do you think we should also have a confirmation of `Y/N` when the user uses the `format` flag? It would be a nice security feature in my opinion.

anshgoyalevil commented:
@EshaanAgg 

Yes that might have helped if we have a different command for format.

A user adding the ```format``` flag is already aware of the fact that database would be formatted.

anshgoyalevil commented on package.json:
```diff
@@ -37,6 +38,7 @@
   "homepage": "https://github.com/PalisadoesFoundation/talawa-api#readme",
   "dependencies": {
     "@graphql-tools/schema": "^8.4.0",
+    "@types/yargs": "^17.0.24",
```
@EshaanAgg 

I think you haven't read the issue statement properly. This script not only needs to be executed in the dev environment, but required in production also.

It will be used by the CI/CD pipeline to execute and parse the arguments.

EshaanAgg replied:
Got it.

anshgoyalevil commented on package.json:
```diff
@@ -72,7 +74,8 @@
     "validator": "^13.7.0",
     "winston": "^3.8.0",
-    "xss-clean": "^0.1.1"
+    "xss-clean": "^0.1.1",
+    "yargs": "^17.7.1"
```
Same for this also.

anshgoyalevil commented on vite.config.ts:
```diff
@@ -30,6 +30,8 @@ export default defineConfig({
         "src/server.ts",
         "src/**/index.ts",
+        "src/utilities/sample_data/*",
+        "src/utilities/loadSampleData.ts",
```
This is a CLI utility, and testing for it is redundant.

EshaanAgg replied:
@anshgoyalevil 
It's not the CLI utility we should be testing but rather the logic of the functions that is included in the same. 
Like for example, is the `formatDatabase` function actually resetting the database or not, do the files for all the models that we are using to populate the database exist, etc.

The goal shouldn't be code coverage, but rather making sure that if in future if we add support for more models, to decide to rework the implementation of these functions, the same does not break the existing logic.

anshgoyalevil replied:
@EshaanAgg 

Your concerns are valid, but take the example of Server.ts, are we testing it? No, because some functions and features shouldn't be tested. For example, we are creating an apollo server in server.ts, but would you go and test if the function is actually being returning a apollo server.

Similarly, we are working with FS in the loadSampleData.ts file.

We would need to mock fs, and yargs for the tests to work. We are acceesing the external functions which cannot be tested. 

I would write the test for it, but, I don't think it would help in any way.

anshgoyalevil commented:
> @palisadoes
> 
> I have updated the `installation.md` also.
> 
> For the automatic database cleanup and import, There are cloud functions available which trigger a script at regular interval of time.
> 
> For example, in Azure, we have [WebJobs](https://learn.microsoft.com/en-us/azure/app-service/webjobs-create) In AWS, we have [AWS Lambda with CloudWatch](https://www.freecodecamp.org/news/using-lambda-functions-as-cronjobs/)

@palisadoes 

Please review.

palisadoes commented on INSTALLATION.md:
```diff
@@ -406,6 +407,53 @@ The key generated in the previous step is in a format suitable for use in a mobi
+| Email                 | Password | User Type  | Joined Organization | Admin For           |
+|-----------------------|----------|------------|---------------------|---------------------|
+| testuser1@gmail.com   | Pass@123 | USER       | Test Organization 2 | None                |
+| testuser2@gmail.com   | Pass@123 | USER       | Test Organization 2 | None                |
```
Use `example.org` instead of `gmail.com` for purposes of copyright.

palisadoes commented on sample_data/events.json:
```diff
@@ -0,0 +1,32 @@
+  {
+    "_id": "6437907585008f171cf2994c",
+    "recurring": true,
+    "recurrance": "ONCE",
```
Make all the events recurring. That way people will always see the events in the default calendar views which would be the current day, month, week.

palisadoes reviewed (changes requested):
See comments.

Also add sample_data directory to the `.github/workflows` authorized files YAML file so that we are always alerted to changes in this critical directory.

anshgoyalevil commented:
Sure Sir. Will make the changes related to above conversations.

palisadoes commented on sample_data/users.json:
```diff
@@ -0,0 +1,177 @@
+    "status": "ACTIVE",
+    "pluginCreationAllowed": true,
+    "adminApproved": true,
+    "firstName": "Test",
```
Please give all the users real sounding names. This will be required if we launch a live demo. We need to start thinking this way from early. There should be random first and last name generators online.

palisadoes commented on sample_data/organizations.json:
```diff
@@ -0,0 +1,87 @@
+    "pinnedPosts": [],
+    "membershipRequests": [],
+    "blockedUsers": [],
+    "name": "Test Organization 1",
```
1. Give the organizations real sounding names, descriptions and locations.
2. Don't use religious names as that may connote bias in a sensitive area. Don't use Palisadoes either.
3. You can end the names with `Charities`, `Foundation` etc. to make them universal.

palisadoes replied:
Do this for all organizations.

palisadoes reviewed (changes requested):
See additional comments

anshgoyalevil commented:
Sure Sir, will make the changes, according to what industry leaders use. Like, will implement what BreezeCHMS use when we try to test a demo of their platform.

anshgoyalevil commented:
@palisadoes 

Sir, I have made the changes.

Gave real names to Users and Organizations, also updated the location.

Changed ```gmail.com``` reference to ```example.com```.

For the events, there is no recurring logic in the API layer, so, I haven't changed that field in the database.
